clang-tools-extra/test/clang-tidy/checkers/readability/redundant-smartptr-get.cpp

```diff
@@ -235,3 +276,13 @@ void Negative() {
   if (MACRO(x) == nullptr)
     ;
 }
+
+void test_redundant_get() {
+  std::vector<std::shared_ptr<int>> v;
+  auto f = [](int) {};
+  for (auto i = v.begin(); i != v.end(); ++i) {
+    f(*i->get());
```

Comment:
try also do a test with something like: f(*i->get()->something())

Reply:
Can I add a member to the shared pointer to achieve this?

Reply:
I think he is saying have a shared_ptr to a class type with a member that can be accessed

Reply:
I tried something like this but I am getting error
#include <memory>
#include <vector>

struct Inner {
  int getValue() const { return 42; }
};

struct Example {
  Inner inner;
  Inner* get() { return &inner; }
};

void test_redundant_get() {
  std::vector<std::shared_ptr<int>> v;
  auto f = [](int) {};
  for (auto i = v.begin(); i != v.end(); ++i) {
    f(**i);
  }
}

void test_redundant_get_with_member() {
  std::vector<std::shared_ptr<Example>> v;
  auto f = [](int) {};
  for (auto i = v.begin(); i != v.end(); ++i) {
    f(*i->get()->getValue());
  }
}

/home/akumar/llvm/a.cpp:29:18: error: no member named 'getValue' in 'Example' [clang-diagnostic-error]
   29 |     f(*i->get()->getValue());
      |        ~~~~~~~~  ^
      
      
  I am unable to get what wrong I am doing here

Reply:
add method getValue in example ... or use get()

Reply:
can you give example please. I tried adding a method

Reply:
`f(i->get()->get()->getValue());`

Reply:
Try with:
f((*i).get()->get()->getValue());

Reply:
It does throw the warning now, but the line remains same, it is not changed

Reply:
I am not  sure if it needs to be changed

Reply:
tried with that, I was able  to reproduce the warning but getting error in lambda.
https://godbolt.org/z/je8f637vb

Reply:
> Trying to analyse the error https://godbolt.org/z/je8f637vb

`f((*i).get()->get()->getValue());` -> `f(*(*i).get()->get()->getValue());`
The `getValue` function returns an `int*`, and not an `int` like the `f` lambda expects. When dereferencing the ptr the code compiles. So either change the return type to a non-ptr `int`, or dereference it to fix the compile-error.

https://godbolt.org/z/vffP16aao

Reply:
thanks for the suggestion. I understand now

Reply:
@PiotrZSL I have raised new pull request since rebased master by mistake and there are log of commits rebased. I will close this once you approve the new pull request.
New pull request: https://github.com/llvm/llvm-project/pull/100177